core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -857,6 +858,364 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
+        status,
+    TeamHandle const& team, int count) {
+  return ThreadVectorRange(team, count);
+}
+
+template <typename TeamHandle>
+KOKKOS_INLINE_FUNCTION auto nested_policy(
+    MDTeamRangeStatus<MDTeamRangeLastRank::NotLastRank,
+                      MDTeamRangeParThread::ParThread,
+                      MDTeamRangeParVector::ParVector>
+        status,
+    TeamHandle const& team, int count) {
+  return TeamVectorRange(team, count);
+}
+
+template <int CurrentRank, int TotalRank, int ParThreadRank, int ParVectorRank,
+          Iterate Iter, typename Policy, typename Lambda, typename... Args>
+KOKKOS_INLINE_FUNCTION void nested_loop(
+    MDTeamRangeStatus<MDTeamRangeLastRank::LastRank,
+                      MDTeamRangeParThread::NotParThread,
+                      MDTeamRangeParVector::NotParVector> const& status,
+    MDTeamRangeNextStatus<CurrentRank, TotalRank, ParThreadRank, ParVectorRank,
+                          Iter> const& nextStatus,
+    Policy const& policy, Lambda const& lambda, Impl::NoReductionTag,
+    Args... args) {
+  lambda(args...);
+}
+
+template <int CurrentRank, int TotalRank, int ParThreadRank, int ParVectorRank,
+          Iterate Iter, typename Policy, typename Lambda,
+          typename ReducerValueType, typename... Args>
+KOKKOS_INLINE_FUNCTION void nested_loop(
+    MDTeamRangeStatus<MDTeamRangeLastRank::LastRank,
+                      MDTeamRangeParThread::NotParThread,
+                      MDTeamRangeParVector::NotParVector> const&,
+    MDTeamRangeNextStatus<CurrentRank, TotalRank, ParThreadRank, ParVectorRank,
+                          Iter> const& nextStatus,
+    Policy const& policy, Lambda const& lambda, ReducerValueType& val,
+    Args... args) {
+  lambda(args..., val);
+}
+
+template <MDTeamRangeParThread ParThread, MDTeamRangeParVector ParVector,
+          int CurrentRank, int TotalRank, int ParThreadRank, int ParVectorRank,
+          Iterate Iter, typename Policy, typename Lambda,
+          typename ReducerValueType, typename... Args>
+KOKKOS_INLINE_FUNCTION void nested_loop(
+    MDTeamRangeStatus<MDTeamRangeLastRank::NotLastRank, ParThread,
+                      ParVector> const& status,
+    MDTeamRangeNextStatus<CurrentRank, TotalRank, ParThreadRank, ParVectorRank,
+                          Iter>,
+    Policy const& policy, Lambda const& lambda, ReducerValueType& val,
+    Args... args) {
+  constexpr int nextRank = CurrentRank + (Iter == Iterate::Right ? 1 : -1);
+  using NextStatus = MDTeamRangeNextStatus<nextRank, TotalRank, ParThreadRank,
+                                           ParVectorRank, Iter>;
+
+  // FIXME Check if recursively calling this is efficient enough compared to an
+  // iterative case
+
+  // This recursively processes ranks from [0..TotalRank-1]
+  // args... is passed by value because it should always be ints
+  parallel_for(
+      nested_policy(status, policy.team, policy.boundaries[CurrentRank]),
```

Comment:
This can't possibly work right now or? I think you need another overload of nested_loop for the case where you have no parallelism.

Reply:
Fixed in [4d802ec](https://github.com/kokkos/kokkos/pull/5238/commits/4d802ec0db4591eda75b5565b0042ef04763c875)